frontend/apps/reader/modules/readersearch.lua

```diff
@@ -51,33 +101,114 @@ function ReaderSearch:onShowFulltextSearchInput()
+                                error_message = error_message .. ":\n" .. SRELL_ERROR_CODES[regex_error]
+                            else
+                                error_message = error_message .. "."
+                            end
+                            UIManager:show(InfoMessage:new{ text = error_message })
+                        else
+                            UIManager:close(self.input_dialog)
+                            self:onShowSearchDialog(self.input_dialog:getInputText(), 1, self.use_regex, self.case_insensitive)
+                        end
                     end,
                 },
                 {
                     text = forward_text,
                     is_enter_default = true,
                     callback = function()
                         if self.input_dialog:getInputText() == "" then return end
-                        UIManager:close(self.input_dialog)
-                        self:onShowSearchDialog(self.input_dialog:getInputText(), 0)
+                        self.last_search_text = self.input_dialog:getInputText()
+                        self.use_regex = self.check_button_regex.checked
+                        self.case_insensitive = not self.check_button_case.checked
+                        local regex_error = self.ui.document:checkRegex(self.input_dialog:getInputText())
+                        if self.use_regex and regex_error ~= 0 then
+                            logger.dbg("ReaderSearch: regex error", regex_error, SRELL_ERROR_CODES[regex_error])
+                            local error_message = _("Invalid regular expression")
+                            if SRELL_ERROR_CODES[regex_error] then
+                                error_message = error_message .. ":\n" .. SRELL_ERROR_CODES[regex_error]
+                            else
+                                error_message = error_message .. "."
+                            end
+                            UIManager:show(InfoMessage:new{ text = error_message })
+                        else
+                            UIManager:close(self.input_dialog)
+                            self:onShowSearchDialog(self.input_dialog:getInputText(), 0, self.use_regex, self.case_insensitive)
+                        end
                     end,
                 },
             },
         },
     }
+
+   -- checkboxes
+    self.check_button_regex = CheckButton:new{
+        text = _("Regular expression (hold for help)"),
+        checked = self.use_regex,
+        parent = self.input_dialog,
+        callback = function()
+            if not self.check_button_regex.checked then
+                self.check_button_regex:check()
+            else
+                self.check_button_regex:unCheck()
+            end
+        end,
+        hold_callback = function()
+            UIManager:show(InfoMessage:new{ text = help_text })
+        end,
+    }
+    self.check_button_case = CheckButton:new{
```

Comment:
Define them in the order they are shown (we had you switch the order at some point).